internal/datastore/proxy/strictreplicated.go

```diff
@@ -305,32 +108,96 @@ func (rr *strictReadReplicatedReader) LookupCaveatsWithNames(ctx context.Context
 	return caveats, err
 }
 
-func (rr *strictReadReplicatedReader) QueryRelationships(
+type queryHandler[F any, O any] func(
 	ctx context.Context,
-	filter datastore.RelationshipsFilter,
-	options ...options.QueryOptionsOption,
+	filter F,
+	options ...O,
+) (datastore.RelationshipIterator, error)
+
+func queryRelationships[F any, O any](
+	ctx context.Context,
+	rr *strictReadReplicatedReader,
+	filter F,
+	options []O,
+	handler func(datastore.Reader) queryHandler[F, O],
 ) (datastore.RelationshipIterator, error) {
 	sr := rr.replica.SnapshotReader(rr.rev)
-	relationships, err := sr.QueryRelationships(ctx, filter, options...)
-	if err != nil && errors.As(err, &common.RevisionUnavailableError{}) {
-		log.Trace().Str("revision", rr.rev.String()).Msg("replica does not contain the requested revision, using primary")
-		return rr.primary.SnapshotReader(rr.rev).QueryRelationships(ctx, filter, options...)
+	it, err := handler(sr)(ctx, filter, options...)
+	// Check for a RevisionUnavailableError, which indicates the replica does not contain the requested
+	// revision. In this case, use the primary instead. This may not be returned on this call from
+	// wrapped datastores that defer the actual query until the iterator is used.
+	if err != nil {
+		if errors.As(err, &common.RevisionUnavailableError{}) {
+			log.Trace().Str("revision", rr.rev.String()).Msg("replica does not contain the requested revision, using primary")
+			return handler(rr.primary.SnapshotReader(rr.rev))(ctx, filter, options...)
+		}
+		return nil, err
 	}
-	return relationships, err
+
+	isFirstResult := true
+	requiresFallback := false
+	return func(yield func(tuple.Relationship, error) bool) {
+	replicaLoop:
+		for result, err := range it {
+			if err != nil {
+				// If the RevisionUnavailableError is returned on the first result, we should fallback
+				// to the primary.
+				if errors.As(err, &common.RevisionUnavailableError{}) {
+					if !isFirstResult {
+						yield(tuple.Relationship{}, spiceerrors.MustBugf("RevisionUnavailableError should only be returned on the first result"))
+						return
+					}
+					requiresFallback = true
+					break replicaLoop
+				}
+
+				if !yield(tuple.Relationship{}, err) {
+					return
+				}
+			}
+
+			isFirstResult = false
+			if !yield(result, nil) {
+				return
+			}
```

Comment:
no code coverage here - can you please also add an assertion to the test that checks the happy path? (revision IS in replica)

Reply:
Added